torch/nn/parameter.py

```diff
@@ -44,3 +44,43 @@ def __reduce_ex__(self, proto):
         )
 
     __torch_function__ = _disabled_torch_function_impl
+
+
+class UninitializedParameter(Parameter):
+    r"""A parameter that is not yet initialized for shape inference support.
+
+    Uninitialized Parameters holds empty tensors that can be moved
+    to different devices and change their type so when they are materialized,
+    they do it directly with the right configuration.
+    """
+    def __new__(cls, requires_grad=True):
+        data = torch.Tensor()
+        return torch.Tensor._make_subclass(cls, data, requires_grad)
+
+    def materialize(self, shape, device=None, dtype=None):
+        r"""Create a Parameter with the same properties of the uninitialized one.
+        Given a shape, it materializes a parameter in the same device
+        and with the same `dtype` as the current one or the specified ones in the 
+        arguments
+        Args:
+            shape : (tuple): the shape for the materialized tensor.
+            device (:class:`torch.device`): the desired device of the parameters
+                and buffers in this module. Optional.
+            dtype (:class:`torch.dtype`): the desired floating point type of
+                the floating point parameters and buffers in this module. Optional.
+        """
+        if device is None:
+            device = self.data.device
+        if dtype is None:
+            dtype = self.data.dtype
+        self.data = torch.empty(shape, device=device, dtype=dtype)
+        self.__class__ = Parameter
```

Comment:
This allows us to cleanly avoid dummy forward passes, as for example the optimizers hold a reference to a parameter that will be materialized in the forward pass of the training

Reply:
That's a good point, maybe we can just set the tensor as property and any access to it will throw an error :)

Reply:
@dzhulgakov I don't think we should expect an optimizer that takes `.data_ptr` to work.
Even a `model.to("cuda")` would break such optimizer.

Reply:
Fair point about data_ptr.

There's actually a bigger problem that optimizers with additional data would initialize in the constructor and try to access the shape of the parameter: https://github.com/pytorch/pytorch/blob/master/torch/optim/adagrad.py#L44 . It might not be a huge deal, because if I read the code correctly - it doesn't work with `.to("cuda")` even today (i.e. one has to move the model before creating the optimizer). But it'd be nice to make `shape` on the parameter to throw before it's initialized.

Reply:
But when can optimizer be safely initialized in this case? After moving model to a target device and making a first iteration, because before that parameters wouldn't have shapes?

Reply:
Actually, in this case, adagrad is not accessing `shape` on the Uninitialized parameter, `full_like` is implemented directly in ATen and is returning an empty tensor :D.

Reply:
But that's not correct right? because when we try to use that state with the the initialized version afterwards it will fail no?

Reply:
Yes, that's actually what I meant but my comment was poorly redacted. It won't hard error in the constructor, but it will error in the step or produce inconsistent results. I added a explicit check to avoid this.